From 126d1fc43696391ff27a83cbb52aea71f52bd80e Mon Sep 17 00:00:00 2001
From: Marek Marczykowski <marmarek@invisiblethingslab.com>
Date: Sun, 7 Apr 2013 19:48:52 +0200
Subject: [PATCH 06/11] RFC: libxl: special 'stubdom-dm' emulator to use qemu
 in stub domain
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Organization: Invisible Things Lab
Cc: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

Xen have feature of having device model in separate domain (called stub
domain). It used to be enabled by special 'stubdom-dm' device model
path.  Recent xl have separate config option for this feature
(device_model_stubdomain_override), but I'm not sure if it worth
introducing another xen-specific option in general domain XML syntax.

Also use os->cmdline as extra arguments for qemu (for HVM only). This
use of os->cmdline is rather dirty hack, but I haven't idea what better
attribute use for it. Perhaps worth new attribute (in <emulator/>
element?). Anyway os->cmdline was unused for HVM domains.

Signed-off-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
---
 src/libxl/libxl_conf.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/libxl/libxl_conf.c b/src/libxl/libxl_conf.c
index 570c6ff..71f5851 100644
--- a/src/libxl/libxl_conf.c
+++ b/src/libxl/libxl_conf.c
@@ -710,6 +710,18 @@ libxlMakeDomBuildInfo(libxlDriverPrivatePtr driver,
             goto error;
         }
 
+        if (def->emulator && STREQ(def->emulator, "stubdom-dm")) {
+            libxl_defbool_set(&b_info->device_model_stubdomain, 1);
+        }
+
+        if (def->os.cmdline && def->os.cmdline[0]) {
+            b_info->extra_hvm = virStringSplit(def->os.cmdline, " ", 0);
+            if (b_info->extra_hvm == NULL) {
+                virReportOOMError();
+                goto error;
+            }
+        }
+
         /* Disable VNC and SDL until explicitly enabled */
         libxl_defbool_set(&b_info->u.hvm.vnc.enable, 0);
         libxl_defbool_set(&b_info->u.hvm.sdl.enable, 0);
-- 
1.8.3.1

